src/services/document-service.js

```diff
@@ -81,35 +86,52 @@ class DocumentService {
   async getSearchableDocumentsMetadataByTags(searchQuery) {
     const tokens = searchQuery.trim().split(/\s+/);
 
-    const positiveTokens = new Set(tokens
-      .filter(token => !token.startsWith('-'))
-      .filter(token => token.length > 2)
-      .map(token => token.toLowerCase()));
-
-    const negativeTokens = new Set(tokens
-      .filter(token => token.startsWith('-'))
-      .map(token => token.substr(1))
-      .filter(token => token.length > 2)
-      .map(token => token.toLowerCase()));
+    const allPositiveTokens = new Set();
+    const positiveExactTokens = new Set();
+    const positivePartialTokens = new Set();
+    const allNegativeTokens = new Set();
+
+    for (const token of tokens) {
+      const isNegative = token.length > 1 && token.startsWith('-');
+      const rawToken = token.substr(isNegative ? 1 : 0).toLowerCase();
+      if (isNegative) {
+        allNegativeTokens.add(rawToken);
+      } else if (rawToken.length < DOCUMENT_PARTIAL_SEARCH_THRESHOLD) {
+        allPositiveTokens.add(rawToken);
+        positiveExactTokens.add(rawToken);
+      } else {
```

Comment:
Please bear with my little nitpicking here.
This 3rd case is a little hard to follow in one's mind (or this one's mind at least), you have to do a recap "so not negative, then length >= ..." 😅 
I normally like to avoid cascading `else` blocks as much I try to avoid nested `if` blocks, when possible, but here I think it would make it easier to follow as :
```
else {
     allPositiveTokens.add(rawToken);

     if (rawToken.length < DOCUMENT_PARTIAL_SEARCH_THRESHOLD) {
        positiveExactTokens.add(rawToken);
     } else {
        positivePartialTokens.add(rawToken);
    }
}
````
Oooor... we just set `allPositiveTokens` by joining `positiveExactTokens` and `positivePartialTokens` at the end of the iteration.

Reply:
Good idea. I followed your first suggestion and divided the logic into nested binary decisions.